### PR TITLE
Use distribution's ValidateRepositoryName for remote name validation. 

### PR DESCRIPTION
**Pull request by shishir-a412ed**
Signed-off-by: Shishir Mahajan shishir.mahajan@redhat.com

### CODE REVIEW

**Comment by shishir-a412ed** [comment]
Follow up on #11116.

**Comment by shishir-a412ed** [comment]
ping @dmcgowan 
any updates on this one ?

**Comment by rhatdan** [comment]
@shishir-a412ed Could you rebase.

**Comment by shishir-a412ed** [comment]
@rhatdan rebased.

File: registry/config.go [hunk 1]
```diff
@@ -206,40 +204,20 @@ func ValidateIndexName(val string) (string, error) {
...
-	}
-	if !validRepo.MatchString(name) {
-		return fmt.Errorf("Invalid repository name (%s), only [a-z0-9-_.] are allowed", name)
+
+	if err := v2.ValidateRespositoryName(remoteName); err != nil {
+		return err
 	}
-	if strings.HasPrefix(name, "-") || strings.HasSuffix(name, "-") {
-		return fmt.Errorf("Invalid repository name (%s). Cannot begin or end with a hyphen.", name)
+	if strings.HasPrefix(remoteName, "-") || strings.HasSuffix(remoteName, "-") {
```
Comment (stevvooe) [inline registry/config.go]: This should be covered by `ValidateRepositoryName`. Do the tests not pass without this condition?

@dmcgowan

File: registry/config.go [hunk 2]
```diff
@@ -206,40 +204,20 @@ func ValidateIndexName(val string) (string, error) {
...
-	if strings.HasPrefix(namespace, "-") || strings.HasSuffix(namespace, "-") {
-		return fmt.Errorf("Invalid namespace name (%s). Cannot begin or end with a hyphen.", namespace)
-	}
-	if strings.Contains(namespace, "--") {
-		return fmt.Errorf("Invalid namespace name (%s). Cannot contain consecutive hyphens.", namespace)
-	}
-	if !validRepo.MatchString(name) {
-		return fmt.Errorf("Invalid repository name (%s), only [a-z0-9-_.] are allowed", name)
+
+	if err := v2.ValidateRespositoryName(remoteName); err != nil {
```
Comment (dmcgowan) [inline registry/config.go]: The type in this name was fixed in the latest version of distribution which should be vendored in master `ValidateRespositoryName -> ValidateRepositoryName`

File: registry/config.go (hunk 1 above)
Comment (dmcgowan) [inline registry/config.go]: Confirmed tests pass without this block

Comment (shishir-a412ed) [inline registry/config.go]: @stevvooe I think the `begin with hyphen` condition is covered in `ValidateRepositoryName`.
Not sure if `end with hyphen` is also covered.

Comment (shishir-a412ed) [inline registry/config.go]: @dmcgowan removed the test condition.

File: registry/config.go (hunk 2 above)
Comment (shishir-a412ed) [inline registry/config.go]: @dmcgowan fixed.

**Comment by dmcgowan** [comment]
Code LGTM

Although I still think we should enumerate what the differences are in validation caused by this change. The validate function will be enforcing this grammar.
```
// 	alpha-numeric := /[a-z0-9]+/
//	separator := /[._-]/
//	component := alpha-numeric [separator alpha-numeric]*
//	namespace := component ['/' component]*
```

**Comment by stevvooe** [comment]
Code LGTM, as well.

**Comment by shishir-a412ed** [comment]
@stevvooe 
'FAIL: docker_cli_tag_test.go:114: DockerSuite.TestTagWithSuffixHyphen

docker_cli_tag_test.go:122:
    c.Fatal("tag a name begin with '-' should failed")
... Error: tag a name begin with '-' should failed
'
Jenky failed with this error. I think this was the reason why I added that extra check in `registry/config.go` earlier. Do you feel we should put the check back ?

Shishir

**Comment by stevvooe** [comment]
@shishir-a412ed I added the exact same test cases to the distribution source and they all pass. Looking at the failure, it seems to be happening in a tagging test. There may be a bug elsewhere.

**Comment by stevvooe** [comment]
@shishir-a412ed I added a few more test cases in distribution: docker/distribution#596.

**Comment by thaJeztah** [comment]
@shishir-a412ed I see you pushed changes to this PR; did you update the code for that failing test, because I see it's still failing.

**Comment by shishir-a412ed** [comment]
@thaJeztah Please check now.

**Comment by thaJeztah** [comment]
Thanks @shishir-a412ed I'll restart the experimental build once more.

ping @stevvooe @dmcgowan PTAL

**Comment by dmcgowan** [comment]
update looks good, tests passing

LGTM

File: integration-cli/docker_cli_tag_test.go [hunk 3]
```diff
@@ -118,13 +118,13 @@ func (s *DockerSuite) TestTagWithSuffixHyphen(c *check.C) {
 	// test repository name begin with '-'
 	tagCmd := exec.Command(dockerBinary, "tag", "busybox:latest", "-busybox:test")
 	out, _, err := runCommandWithOutput(tagCmd)
-	if err == nil || !strings.Contains(out, "Invalid repository name (-busybox). Cannot begin or end with a hyphen") {
+	if err == nil || !strings.Contains(out, "repository name component must match") {
```
Comment (thaJeztah) [inline integration-cli/docker_cli_tag_test.go]: Just noticed this test is called `TestTagWithSuffixHyphen`, but we're testing a *prefix*, not a *suffix*
(Not introduced in this PR)

Comment (shishir-a412ed) [inline integration-cli/docker_cli_tag_test.go]: @thaJeztah Changed it to `TestTagWithPrefixHyphen`

Comment (thaJeztah) [inline integration-cli/docker_cli_tag_test.go]: :+1: thanks!

**Comment by dmcgowan** [comment]
ping @tibor @RichardScothern

**Comment by RichardScothern** [comment]
Tidy.  LGTM

**Comment by shishir-a412ed** [comment]
@tiborvass any updates on this one ?

**Comment by tiborvass** [comment]
@shishir-a412ed what is the purpose of this PR? To allow `/foo/bar/image` ? Is there any other change?

**Comment by shishir-a412ed** [comment]
@tiborvass, yeah and using the distribution's validation for both local and registry.

File: registry/config.go [hunk 4]
```diff
@@ -206,41 +204,19 @@ func ValidateIndexName(val string) (string, error) {
...
-	if strings.Contains(namespace, "--") {
-		return fmt.Errorf("Invalid namespace name (%s). Cannot contain consecutive hyphens.", namespace)
-	}
-	if !validRepo.MatchString(name) {
-		return fmt.Errorf("Invalid repository name (%s), only [a-z0-9-_.] are allowed", name)
-	}
-	if strings.HasPrefix(name, "-") || strings.HasSuffix(name, "-") {
-		return fmt.Errorf("Invalid repository name (%s). Cannot begin or end with a hyphen.", name)
+
+	if err := v2.ValidateRepositoryName(remoteName); err != nil {
```
Comment (tiborvass) [inline registry/config.go]: @shishir-a412ed `return v2.ValidateRepositoryName(...`

Comment (shishir-a412ed) [inline registry/config.go]: @tiborvass changed.

**Comment by tiborvass** [comment]
LGTM

**Comment by dmcgowan** [comment]
LGTM

**Comment by thaJeztah** [comment]
Thanks @shishir-a412ed !

**Comment by rhatdan** [comment]
woohoo

**Pull request merged**